src/coreclr/jit/assertionprop.cpp

```diff
@@ -1323,6 +1323,7 @@ AssertionIndex Compiler::optCreateAssertion(GenTree*         op1,
         //
         // Set op1 to the instance pointer of the indirection
         //
+        op1 = op1->gtEffectiveVal(/* commaOnly */ true);
```

Comment:
Nice find! I see `optAssertionProp_Cast` calls `gtEffectiveVal` without commaOnly argument in a similar case - is it needed? (not sure I understand when we can face a GT_NOP with a child non-null node)

Reply:
It's a throughput optimization to use `commaOnly` in this case, it halved the regression I was seeing with PIN.

The general IR contract is that `NOP`s can be used as "passthrough" for non-`TYP_STRUCT` values, so `gtEffectiveVal(/* commaOnly */ false)` will catch more cases. That said, there is only one place which generates such `NOP`s (`optNarrowTree`), so I would consider using `/* commaOnly */ true` to have better CQ/TP value.

I was actually considering deleting the `NOP` handling altogether but it of course doesn't belong in this change.